### ISSUE TITLE
@JSExport on method with variable arguments in trait behaves different from class

### ISSUE THREAD

**Issue by joschuafink** [scala-js/scala-js#3538]
In Scala.js 0.6.26 the from JavaScript accessible method `possiblyBroken` throws an Exception when being called from JavaScript. A look in the compiled code shows, that it lacks the code for handling variable arguments that `notBroken1` and `notBroken2` have.

```scala
trait Foo {
  @JSExport def possiblyBroken(args: Int*): Unit = ???
}

class Foo2 {
  @JSExport def notBroken1(args: Int*): Unit = ???
}

@JSExportTopLevel("bar")
object Bar extends Foo2 with Foo {
  @JSExport def notBroken2(args: Int*): Unit = ???
}
```

**Comment by sjrd** [comment]
Thank you for the report. I can reproduce. Definitely a bug.

**Comment by sjrd** [comment]
So, this is going to be very tricky :-s

Normally, to detect whether a parameter was a vararg, once we're in the Scala.js compiler back-end, we need to use time travel in the type history of scalac, all the way back to before `uncurry`. The problem is that for a method in a trait, scalac creates a bridge for `broken` in `Bar` during `mixin`, and the bridge *doesn't* have the type history we need. And that's what's causing the bug.

Funnily enough, it seems that scalac already has some code to copy the type history up to entering-erasure, but not further:
https://github.com/scala/scala/blob/e40c95e6e51b215c8cbb0b9e438fa7a5d5ed8332/src/compiler/scala/tools/nsc/transform/Mixin.scala#L155
It needs that to generate the proper generic signatures for Java, apparently: https://github.com/scala/bug/issues/3857

I have explored a few ways to work around the lack of information on our side, but I am hitting dead ends on all sides. I believe that the easiest fix would be to amend scalac to copy the type history all the way back to entering-uncurry. That would mean that the bug in Scala.js would only be fixed in newer versions of Scala (whereas usually we fix bugs all the way to Scala 2.10.2).

**Comment by sjrd** [comment]
Fixed in 08a3c1497855d949a13af6fe367ea3f1c7211cf1.
